content/blog/updating-to-cpp11/index.Rmd

```diff
@@ -0,0 +1,61 @@
+---
+output: hugodown::hugo_document
+
+slug: updating-to-cpp11
+title: Pathway to success - updating your package to cpp11
+date: 2021-09-10
+author: Shelby Bearrows
+description: >
+    The cpp11 summer intern reviews the process they used to convert readxl to using 
+    cpp11.
+
+photo:
+  url: https://unsplash.com/photos/9MMd2uRpfvc
+  author: John Salzarulo
+
+# one of: "deep-dive", "learn", "package", "programming", or "other"
+categories: [learn] 
+tags: [cpp11, internship]
+---
+
+<!--
+TODO:
+* [ ] Look over / edit the post's title in the yaml
+* [ ] Edit (or delete) the description; note this appears in the Twitter card
+* [ ] Pick category and tags (see existing with `hugodown::tidy_show_meta()`)
+* [ ] Find photo & update yaml metadata
+* [x] Create `thumbnail-sq.jpg`; height and width should be equal
+* [x] Create `thumbnail-wd.jpg`; width should be >5x height
+* [x] `hugodown::use_tidy_thumbnails()`
+* [ ] Add intro sentence, e.g. the standard tagline for the package
+* [ ] `usethis::use_tidy_thanks()`
+-->
+
+Over the summer I had the pleasure of working with Jim Hester on the cpp11 package as a tidyverse summer intern. The cpp11 package is a header-only R package that helps R package developers handle R objects with C++ code. Its goals and syntax are similar to the excellent Rcpp package.
+
+During most of my internship, I worked on triaging issue requests, fixing bugs and adding new features to cpp11. Near the end of the summer, I got to work with Jenny Bryan on converting the readxl package from using Rcpp to using cpp11.
+
+Jim has written a [great post](https://cpp11.r-lib.org/articles/converting.html) about converting packages from Rcpp to cpp11 which I heavily referenced during the process . But there were still some challenges I encountered. To help others going through a similar process, I wanted to review the workflows and tools we used to make this easier.
+
+## Where to start?
+ 
+To get started, I followed the initial set-up steps outlined in Jim’s article and then recompiled, to confirm it was successful. Next, I needed to include either the cpp11/R.hpp header or the macros R_NO_REMAP and STRICT_R_HEADERS. I found it easier to include the macros since the header file needs to come before any headers that use Rcpp. As for which file to start with, I’d suggest starting from any header or C++ file that is the most self-contained. Initially, I made the mistake of selecting the most interconnected file in the readxl package! Since this file was central to the package, I would have had to convert the entire package over to cpp11 before the compilation errors would stop. After Jim and Jenny helped me select a more approachable file, the process was more incremental and I successfully maintained my sanity 🙌🏼.
```

Comment:
```suggestion
To get started, I followed the initial set-up steps outlined in Jim’s article and then recompiled, to confirm it was successful. Next, I needed to include either the `cpp11/R.hpp` header or the macros `R_NO_REMAP` and `STRICT_R_HEADERS`. I found it easier to include the macros since the header file needs to come before any headers that use Rcpp. As for which file to start with, I’d suggest starting from any header or C++ file that is the most self-contained. Initially, I made the mistake of selecting the most interconnected file in the readxl package! Since this file was central to the package, I would have had to convert the entire package over to cpp11 before the compilation errors would stop. After Jim and Jenny helped me select a more approachable file, the process was more incremental and I successfully maintained my sanity 🙌🏼.
```

This section / paragraph implicitly sends the message "we're going to approach this task incrementally, one file at a time", but I think it would be nice to say that explicitly at that start. Maybe the section header should also change accordingly.

I think this mindset, and the idea of moving file-by-file, keeping everything "green" (in the "tests pass" sense), is a big nugget of wisdom in this post. You might also want to mention the points at which you would make a commit and push (and get full CI).

Reply:
Love this! Great idea